scheduler/scheduler.go

```diff
@@ -1,122 +1,328 @@
+		QueuedRequests: st.QueuedJobs,
 	}
 }
 
-func (s *Stack) Wait() (done func(), err error) {
-	return s.stack.Wait()
+// Config returns the configuration that the queue was created with.
+func (q *Queue) Config() Config {
+	return q.config
 }
 
-func (s *Stack) close() {
-	s.stack.Close()
+func (q *Queue) reconfigure() {
+	// renaming Stack -> Queue in the jobqueue project will follow
+	q.queue.Reconfigure(jobqueue.Options{
+		MaxConcurrency: q.config.MaxConcurrency,
+		MaxStackSize:   q.config.MaxQueueSize,
+		Timeout:        q.config.Timeout,
+	})
 }
 
-func (s *Stack) Config() Config {
-	return s.config
+func (q *Queue) close() {
+	q.queue.Close()
 }
 
-func NewRegistry() *Registry {
+// RegistryWith (Options) creates a registry with the provided options.
+func RegistryWith(o Options) *Registry {
+	if o.MetricsUpdateTimeout <= 0 {
+		o.MetricsUpdateTimeout = time.Second
+	}
+
 	return &Registry{
-		groupConfig: make(map[string]Config),
-		stacks:      make(map[string]*Stack),
+		options: o,
+		queues:  new(sync.Map),
+		quit:    make(chan struct{}),
 	}
 }
 
-func (r *Registry) Config(key string) Config {
-	r.mu.Lock()
-	cfg := r.groupConfig[key]
-	r.mu.Unlock()
-	return cfg
+// NewRegistry creates a registry with the default options.
+func NewRegistry() *Registry {
+	return RegistryWith(Options{})
 }
 
-func (r *Registry) getStack(name string) (s *Stack, ok bool) {
-	s, ok = r.stacks[name]
-	return
-}
+func (r *Registry) newQueue(name string, c Config) *Queue {
+	measure := r.options.EnableRouteLIFOMetrics
+	q := &Queue{
+		config: c,
+		// renaming Stack -> Queue in the jobqueue project will follow
+		queue: jobqueue.With(jobqueue.Options{
+			MaxConcurrency: c.MaxConcurrency,
+			MaxStackSize:   c.MaxQueueSize,
+			Timeout:        c.Timeout,
+		}),
+	}
 
-func (r *Registry) setStack(name string, s *Stack) {
-	r.stacks[name] = s
+	if measure {
+		if name == "" {
+			name = "unknown"
+		}
+
+		q.activeRequestsMetricsKey = fmt.Sprintf("lifo.%s.active", name)
+		q.queuedRequestsMetricsKey = fmt.Sprintf("lifo.%s.queued", name)
+		r.measure()
+	}
+
+	return q
 }
 
-// Do implements routing.PostProcessor and sets the stack for the scheduler filters.
-//
-// It preserves the existing stack when available.
-func (r *Registry) Do(routes []*routing.Route) []*routing.Route {
+func (r *Registry) initLIFOFilters(routes []*routing.Route) []*routing.Route {
 	rr := make([]*routing.Route, len(routes))
+	existingKeys := make(map[string]bool)
+	groups := make(map[string][]GroupedLIFOFilter)
+
 	for i, ri := range routes {
 		rr[i] = ri
 		for _, fi := range ri.Filters {
+			// TODO: warn on multiple lifos in the same route
+			if glf, ok := fi.Filter.(GroupedLIFOFilter); ok {
+				groupName := glf.Group()
+				groups[groupName] = append(groups[groupName], glf)
+				continue
+			}
+
 			lf, ok := fi.Filter.(LIFOFilter)
+			if !ok {
+				continue
+			}
+
+			var q *Queue
+			key := fmt.Sprintf("lifo::%s", ri.Id)
+			existingKeys[key] = true
```

Comment:
PostProcessors are not running concurrently, but the tests could fail, if we run with race detector, which we do.
Maybe change to sync.Map ?

Reply:
Based on my understanding about how the race detector works, which may not be complete, this is not a problem. I tested with -race, and this here seems to be fine. Let's fix it once it is a problem.